model/Image.php

```diff
@@ -26,60 +27,106 @@ public function getIsImage() {
 	}
 
 	/**
-	 * Helper method to regenerate all image links in the given HTML block, optionally resizing them if
-	 * the image native size differs to the width and height properties on the <img /> tag
+	 * Replace"[image id=n]" shortcode with an image reference.
+	 * Permission checks will be enforced by the file routing itself.
 	 *
-	 * @param string $value HTML value
-	 * @return string value with links resampled
+	 * @param array $args Arguments passed to the parser
+	 * @param string $content Raw shortcode
+	 * @param ShortcodeParser $parser Parser
+	 * @param string $shortcode Name of shortcode used to register this handler
+	 * @param array $extra Extra arguments
+	 * @return string Result of the handled shortcode
 	 */
-	public static function regenerate_html_links($value) {
-		$htmlValue = Injector::inst()->create('HTMLValue', $value);
-
-		// Resample images and add default attributes
-		$imageElements = $htmlValue->getElementsByTagName('img');
-		if($imageElements) foreach($imageElements as $imageElement) {
-			$imageDO = null;
-			$src = $imageElement->getAttribute('src');
+	public static function handle_shortcode($args, $content, $parser, $shortcode, $extra = array()) {
+		// Check if there is a suitable record
+		$record = static::find_shortcode_record($args);
+		if (!$record) {
+			return null;
+		}
```

Comment:
You're missing an important canView() check before the `$record->Link()`... this will potentially expose non-allowed assets to users or site visitors.